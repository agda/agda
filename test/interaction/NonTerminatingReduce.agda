-- Testcase by Ulf (2014-07-07, commit c8128d1).
-- Revised by Andreas and Ulf 2025-09-16 for issue #2410.

open import Agda.Builtin.Nat
open import Agda.Builtin.Bool

if_then_else_ : {A : Set} → Bool → A → A → A
if true  then a else b = a
if false then a else b = b

pred : Nat → Nat
pred zero = zero
pred (suc n) = n

-- The following function diverging for unknown n, so we need to label it NON_TERMINATING.
-- At run-time it terminates, so we might want to evaluate it on some concrete numbers.
{-# NON_TERMINATING #-}
loop : Nat → Nat
loop n = if n == 0 then 0 else loop (pred n)

hole : Set
hole = {!!}

-- Non-terminating functions do not reduce by default (C-c C-n),
-- only when more force is applied (C-u C-c C-n).
