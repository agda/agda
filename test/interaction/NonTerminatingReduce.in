top_command (cmd_load currentFile [])

-- Should be 'loop 5'
top_command (cmd_compute_toplevel DefaultCompute "loop 5")

-- Should be 'loop 5'
goal_command 0 (cmd_compute DefaultCompute) "loop 5"

-- Should be '0'
top_command (cmd_compute_toplevel IgnoreAbstract "loop 5")

-- Should be '0'
goal_command 0 (cmd_compute IgnoreAbstract) "loop 5"

