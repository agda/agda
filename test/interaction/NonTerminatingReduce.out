(agda2-status-action "")
(agda2-info-action "*Type-checking*" "" nil)
(agda2-highlight-clear)
(agda2-status-action "")
(agda2-info-action "*All Goals*" "?0 : Set" nil)
((last . 1) . (agda2-goals-action '(0)))
(agda2-status-action "")
(agda2-info-action "*Normal Form*" "loop 5" nil)
(agda2-status-action "")
(agda2-info-action "*Normal Form*" "loop 5" nil)
(agda2-status-action "")
(agda2-info-action "*Normal Form*" "0" nil)
(agda2-status-action "")
(agda2-info-action "*Normal Form*" "0" nil)
